protos/local/failure.proto

```diff
@@ -0,0 +1,77 @@
+syntax = "proto3";
+
+package coresdk.failure;
+
+import "google/protobuf/duration.proto";
+import "common.proto";
+
+/// Represents a failure in user code, workflow or activity, which could've been triggered by
+/// an exception or similar error mechanism like the error half of a Result type.
+message ApplicationFailureInfo {
+    string type = 1;
+    bool non_retryable = 2;
+    repeated common.Payload details = 3;
+}
+
+/// Represents a timeout, for example of a child workflow or activity.
+message TimeoutFailureInfo {
+    common.TimeoutType timeout_type = 1;
+    repeated common.Payload last_heartbeat_details = 2;
+}
+
+/// Represents a cancellation, for example of a child workflow or activity.
+message CancelledFailureInfo {
+    repeated common.Payload details = 1;
+}
+
+/// Represents workflow termination.
+message TerminatedFailureInfo {
+}
+
+/// Represents failure on the server side
+message ServerFailureInfo {
+    bool non_retryable = 1;
+}
+
+/// TODO: when does this happen?
+message ResetWorkflowFailureInfo {
+    repeated common.Payload last_heartbeat_details = 1;
+}
+
+/// An activity has failed, TODO: how is this generated?
+message ActivityFailureInfo {
+    int64 scheduled_event_id = 1;
+    int64 started_event_id = 2;
+    string identity = 3;
+    string activity_type = 4;
+    string activity_id = 5;
+    common.RetryState retry_state = 6;
+}
+
+/// Child workflow execution has failed.
+message ChildWorkflowExecutionFailureInfo {
+    string namespace = 1;
+    common.WorkflowExecution workflow_execution = 2;
+    string workflow_type = 3;
+    int64 initiated_event_id = 4;
+    int64 started_event_id = 5;
+    common.RetryState retry_state = 6;
+}
+
+/// Structured failure which can be mapped to upstream proto Failure
+message Failure {
```

Comment:
As discussed in slack this is nasty and we shouldn't copy it. Way too easy to represent impossible states since this `Failure` includes every possible failure type but only some failures are possible depending on what's failing.